src/software/thunderscope/robot_diagnostics/controller_diagnostics.py

```diff
@@ -0,0 +1,245 @@
+import logging
+
+from evdev import InputDevice, categorize, ecodes, list_devices
+from threading import Event, Thread
+
+from proto.import_all_protos import *
+from software.thunderscope.constants import *
+from software.thunderscope.proto_unix_io import ProtoUnixIO
+import software.python_bindings as tbots_cpp
+
+XBOX_MAX_RANGE = 32768
+XBOX_BUTTON_MAX_RANGE = 1024
+
+DEADZONE_PERCENTAGE = 0.30
+
+DRIBBLER_STEPPER = 100
+POWER_STEPPER = 100
+
+MAX_LINEAR_SPEED_METER_PER_S = 2
+MAX_ANGULAR_SPEED_RAD_PER_S = 10.0
+
+
+class ControllerInputHandler(object):
+    """
+    This class is responsible for reading from an Xbox controller device and
+    interpreting the inputs into usable inputs for robot.
+    """
+
+    # TODO: remove proto_unix_io, and set Motor/Power control as class fields
+    # TODO: add class init wrapper for easier handling of controller connection
+    def __init__(
+        self, proto_unix_io: ProtoUnixIO,
+    ):
+        self.proto_unix_io = proto_unix_io
+        self.enabled = False
+        self.controller = None
+
+        # find an
+        for device in list_devices():
+            controller = InputDevice(device)
+            # TODO: read the device name and check that it's xbox controller
+            if controller is not None:
+                self.controller = controller
```

Comment:
It would be nice if this logic was refactored to some helper method which is maybe connected to a "refresh" button in the GUI. There we would also have the status of if a controller is detected and connected or not.

If you do this refactor, the rest of this constructor will need to be updated as well.

Reply:
Can discuss more in meetings...